samples/19.a.single-sign-on-for-enterprise-apps/README.md

```diff
@@ -0,0 +1,195 @@
+1. [Setup OAuth via GitHub](#setup-oauth-via-github)
+1. [Setup Azure Bot Services](#setup-azure-bot-services)
+1. [Prepare and run the code](#prepare-and-run-the-code)
+
+## Clone the code
+
+To host this demo, you will need to clone the code and run locally.
+
+1. Clone this repository
+1. Create two files for environment variables, `/bot/.env` and `/rest-api/.env`
+   - In `/rest-api/.env`
+      - Write `AAD_OAUTH_REDIRECT_URI=http://localhost:3000/api/aad/oauth/callback`
+         - When Azure Active Directory complete the authorization flow, it will send the browser to this URL. This URL must be accessible by the browser from end-user machine
+      - Write `GITHUB_OAUTH_REDIRECT_URI=http://localhost:3000/api/github/oauth/callback`
+         - Same as Azure Active Directory, this is the URL for GitHub to send its result
+
+## Setup OAuth via Azure Active Directory
+
+If you want to authenticate on Azure Active Directory, follow the steps below.
+
+- Go to [your Azure Active Directory](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview)
+- Create a new application
+   1. Select "App registrations"
+   1. Click "New registration"
+   1. Give it a name
+   1. In "Redirect URI (optional)" section, add a new entry
+      1. Select "web" as type
+      1. Enter `http://localhost:3000/api/aad/oauth/callback` as redirect URI
+         - This must match `AAD_OAUTH_REDIRECT_URI` in `/rest-api/.env` we saved earlier
+   - Click "Register"
+- Save the authorization endpoints
+   1. Select "Overview"
+   1. Click "Endpoints" on the main pane
+   1. Copy the content of "OAuth 2.0 authorization endpoint (v2)" to `/rest-api/.env`
+      - `AAD_OAUTH_AUTHORIZE_URL=https://login.microsoftonline.com/12345678-1234-5678-abcd-12345678abcd/oauth2/v2.0/authorize`
+   1. Copy the content of "OAuth 2.0 token endpoint (v2)" to `/rest-api/.env`
+      - `AAD_OAUTH_ACCESS_TOKEN_URL=https://login.microsoftonline.com/12345678-1234-5678-abcd-12345678abcd/oauth2/v2.0/token`
+- Save the client ID
+   1. Select "Overview"
+   1. On the main pane, copy the content of "Application (client) ID" to `/rest-api/.env`
+      - `AAD_OAUTH_CLIENT_ID=12345678abcd-1234-5678-abcd-12345678abcd`
+- Save the client secret
+   1. Select "Certificates & secrets"
+   1. In the "Client secrets" section, create a "New client secret"
+   1. Save the client secret to `/rest-api/.env` file
+      - `AAD_OAUTH_CLIENT_SECRET=<your app client secret>`
+- Enable "Implicit Grant" flow
+   1. Select "Authentication"
+   1. In the "Advanced Settings" section, check "Access tokens"
+   1. Click "Save" button to save the changes
+
+## Setup OAuth via GitHub
+
+If you want to authenticate on GitHub, follow the steps below.
+
+1. Sign into GitHub
+1. Create a new OAuth application
+   1. Browse to https://github.com/settings/developers
+   1. Select "OAuth Apps"
+   1. Click "New OAuth App"
+   1. Fill out "Application name" and "Homepage URL"
```

Comment:
Should homepage URL be localhost:3000 or something else?

Reply:
This URL is for GitHub only and for legal purpose.

When authorizing an app, GitHub will show a page telling users that what organization is developing this OAuth application. And this is the URL which GitHub point to for legal purpose.

For development purpose, user could put anything on, including http://localhost:3000. For production, they are required by GitHub to put a real URL on.

I prefer not to ask the user to fill it out using localhost:3000, fearing that they might forget to set a real one on their production app.

Reply:
Makes sense. Maybe just add a quick comment saying that this URL should point to the user's legal page regarding enabling OAuth.

Reply:
Added.